File: cortex/python/app/components/workspace_manager.py
"""
Workspace Manager Component

This module implements a workspace manager that handles workspaces for user sessions.
Workspaces serve as containers for related activities, data, and context.
"""

import asyncio
import uuid
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Set, Union, Any

from fastapi import HTTPException, status
from pydantic import BaseModel, Field

from app.cache.redis import redis_cache
from app.config import settings
from app.database.connection import get_session
from app.database.models import Workspace as WorkspaceModel
from app.utils.logger import get_contextual_logger

# Configure logger
logger = get_contextual_logger("components.workspace_manager")


class WorkspaceMetadata(BaseModel):
    """Model for workspace metadata"""

    name: Optional[str] = None
    description: Optional[str] = None
    type: str = "default"
    tags: List[str] = Field(default_factory=list)
    custom_data: Dict[str, Any] = Field(default_factory=dict)


class Workspace(BaseModel):
    """Model representing a workspace"""

    id: uuid.UUID = Field(default_factory=uuid.uuid4)
    owner_id: Optional[uuid.UUID] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    last_active_at: datetime = Field(default_factory=datetime.utcnow)
    is_archived: bool = False
    is_private: bool = True
    metadata: WorkspaceMetadata = Field(default_factory=WorkspaceMetadata)

    class Config:
        schema_extra = {
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "owner_id": "550e8400-e29b-41d4-a716-446655440001",
                "created_at": "2025-03-06T15:30:45.123456",
                "updated_at": "2025-03-06T15:30:45.123456",
                "last_active_at": "2025-03-06T15:30:45.123456",
                "is_archived": False,
                "is_private": True,
                "metadata": {
                    "name": "My Workspace",
                    "description": "A workspace for my project",
                    "type": "project",
                    "tags": ["project", "important"],
                    "custom_data": {"priority": "high"},
                },
            }
        }


class WorkspaceManager:
    """
    Workspace Manager for handling user workspaces

    This class provides functionality for creating, retrieving, and managing
    workspaces, which serve as containers for related activities and context.
    """

    WORKSPACE_CACHE_PREFIX = "workspace:"
    WORKSPACE_CACHE_TTL = 3600  # 1 hour in seconds

    def __init__(self):
        """Initialize the workspace manager"""
        # In-memory cache for active workspaces
        self.active_workspaces: Dict[uuid.UUID, Workspace] = {}

        # User workspace mapping (user_id -> set of workspace_ids)
        self.user_workspaces: Dict[uuid.UUID, Set[uuid.UUID]] = {}

        logger.info("Workspace manager initialized")

    async def create_workspace(
        self,
        owner_id: Optional[uuid.UUID] = None,
        name: Optional[str] = None,
        description: Optional[str] = None,
        workspace_type: str = "default",
        is_private: bool = True,
        tags: Optional[List[str]] = None,
        custom_data: Optional[Dict[str, Any]] = None,
    ) -> Workspace:
        """
        Create a new workspace

        Args:
            owner_id: Optional owner user ID
            name: Optional workspace name
            description: Optional workspace description
            workspace_type: Workspace type
            is_private: Whether the workspace is private
            tags: Optional tags for the workspace
            custom_data: Optional custom data for the workspace

        Returns:
            New workspace
        """
        try:
            # Create metadata
            metadata = WorkspaceMetadata(
                name=name,
                description=description,
                type=workspace_type,
                tags=tags or [],
                custom_data=custom_data or {},
            )

            # Create workspace
            workspace_id = uuid.uuid4()
            workspace = Workspace(
                id=workspace_id,
                owner_id=owner_id,
                is_private=is_private,
                metadata=metadata,
            )

            # Save to database
            async with get_session() as db:
                db_workspace = WorkspaceModel(
                    id=str(workspace.id),
                    owner_id=str(workspace.owner_id) if workspace.owner_id else None,
                    created_at=workspace.created_at,
                    updated_at=workspace.updated_at,
                    last_active_at=workspace.last_active_at,
                    is_archived=workspace.is_archived,
                    is_private=workspace.is_private,
                    metadata=workspace.metadata.dict(),
                )

                db.add(db_workspace)
                await db.commit()
                await db.refresh(db_workspace)

            # Add to cache
            self.active_workspaces[workspace_id] = workspace
            await self._cache_workspace(workspace)

            # Update user workspace mapping
            if owner_id:
                if owner_id not in self.user_workspaces:
                    self.user_workspaces[owner_id] = set()
                self.user_workspaces[owner_id].add(workspace_id)

            logger.info(f"Created workspace: {workspace_id}")
            return workspace

        except Exception as e:
            logger.error(f"Failed to create workspace: {str(e)}", exc_info=True)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Failed to create workspace: {str(e)}",
            )

    async def get_workspace(self, workspace_id: uuid.UUID) -> Optional[Workspace]:
        """
        Get a workspace by ID

        Args:
            workspace_id: Workspace ID

        Returns:
            Workspace or None if not found
        """
        try:
            # Check in-memory cache first
            workspace = self.active_workspaces.get(workspace_id)

            if workspace is None:
                # Try to get from Redis cache
                cached_workspace = await redis_cache.get(
                    key=self._get_cache_key(workspace_id)
                )

                if cached_workspace:
                    workspace = Workspace(**cached_workspace)

                    # Add to in-memory cache
                    self.active_workspaces[workspace_id] = workspace

                    # Update user workspace mapping
                    if workspace.owner_id:
                        if workspace.owner_id not in self.user_workspaces:
                            self.user_workspaces[workspace.owner_id] = set()
                        self.user_workspaces[workspace.owner_id].add(workspace_id)

            if workspace is None:
                # Try to get from database
                async with get_session() as db:
                    db_workspace = await db.get(WorkspaceModel, str(workspace_id))

                    if db_workspace:
                        # Convert to Workspace model
                        workspace = Workspace(
                            id=uuid.UUID(db_workspace.id),
                            owner_id=uuid.UUID(db_workspace.owner_id)
                            if db_workspace.owner_id
                            else None,
                            created_at=db_workspace.created_at,
                            updated_at=db_workspace.updated_at,
                            last_active_at=db_workspace.last_active_at,
                            is_archived=db_workspace.is_archived,
                            is_private=db_workspace.is_private,
                            metadata=WorkspaceMetadata(**db_workspace.metadata),
                        )

                        # Add to caches
                        self.active_workspaces[workspace_id] = workspace
                        await self._cache_workspace(workspace)

                        # Update user workspace mapping
                        if workspace.owner_id:
                            if workspace.owner_id not in self.user_workspaces:
                                self.user_workspaces[workspace.owner_id] = set()
                            self.user_workspaces[workspace.owner_id].add(workspace_id)

            if workspace is None:
                logger.warning(f"Workspace not found: {workspace_id}")
                return None

            # Update last active time
            workspace.last_active_at = datetime.utcnow()
            workspace.updated_at = datetime.utcnow()

            # Update cache
            await self._cache_workspace(workspace)

            return workspace

        except Exception as e:
            logger.error(f"Failed to get workspace: {str(e)}", exc_info=True)
            return None

    async def update_workspace(
        self,
        workspace_id: uuid.UUID,
        owner_id: Optional[uuid.UUID] = None,
        name: Optional[str] = None,
        description: Optional[str] = None,
        workspace_type: Optional[str] = None,
        is_private: Optional[bool] = None,
        is_archived: Optional[bool] = None,
        tags: Optional[List[str]] = None,
        custom_data: Optional[Dict[str, Any]] = None,
    ) -> Optional[Workspace]:
        """
        Update a workspace

        Args:
            workspace_id: Workspace ID to update
            owner_id: Optional new owner user ID
            name: Optional new workspace name
            description: Optional new workspace description
            workspace_type: Optional new workspace type
            is_private: Optional new privacy setting
            is_archived: Optional new archive status
            tags: Optional new tags
            custom_data: Optional new custom data

        Returns:
            Updated workspace or None if not found
        """
        try:
            # Get current workspace
            workspace = await self.get_workspace(workspace_id)
            if workspace is None:
                logger.warning(f"Workspace not found for update: {workspace_id}")
                return None

            # Update workspace properties
            if owner_id is not None:
                # If owner changed, update user workspace mapping
                if workspace.owner_id and workspace.owner_id != owner_id:
                    if workspace.owner_id in self.user_workspaces:
                        self.user_workspaces[workspace.owner_id].discard(workspace_id)

                workspace.owner_id = owner_id

                # Add to new owner's workspace list
                if owner_id:
                    if owner_id not in self.user_workspaces:
                        self.user_workspaces[owner_id] = set()
                    self.user_workspaces[owner_id].add(workspace_id)

            if is_private is not None:
                workspace.is_private = is_private

            if is_archived is not None:
                workspace.is_archived = is_archived

            # Update metadata
            metadata_updated = False

            if name is not None:
                workspace.metadata.name = name
                metadata_updated = True

            if description is not None:
                workspace.metadata.description = description
                metadata_updated = True

            if workspace_type is not None:
                workspace.metadata.type = workspace_type
                metadata_updated = True

            if tags is not None:
                workspace.metadata.tags = tags
                metadata_updated = True

            if custom_data is not None:
                workspace.metadata.custom_data = custom_data
                metadata_updated = True

            # Update timestamps
            workspace.updated_at = datetime.utcnow()
            workspace.last_active_at = datetime.utcnow()

            # Save to database
            async with get_session() as db:
                db_workspace = await db.get(WorkspaceModel, str(workspace_id))

                if db_workspace:
                    db_workspace.owner_id = (
                        str(workspace.owner_id) if workspace.owner_id else None
                    )
                    db_workspace.updated_at = workspace.updated_at
                    db_workspace.last_active_at = workspace.last_active_at
                    db_workspace.is_archived = workspace.is_archived
                    db_workspace.is_private = workspace.is_private
                    db_workspace.metadata = workspace.metadata.dict()

                    await db.commit()
                    await db.refresh(db_workspace)
                else:
                    logger.warning(f"Workspace not found in database: {workspace_id}")

            # Update caches
            self.active_workspaces[workspace_id] = workspace
            await self._cache_workspace(workspace)

            logger.info(f"Updated workspace: {workspace_id}")
            return workspace

        except Exception as e:
            logger.error(f"Failed to update workspace: {str(e)}", exc_info=True)
            return None

    async def delete_workspace(self, workspace_id: uuid.UUID) -> bool:
        """
        Delete a workspace

        Args:
            workspace_id: Workspace ID to delete

        Returns:
            True if deleted, False if not found or error
        """
        try:
            # Get current workspace
            workspace = await self.get_workspace(workspace_id)
            if workspace is None:
                logger.warning(f"Workspace not found for deletion: {workspace_id}")
                return False

            # Remove from user workspace mapping
            if workspace.owner_id and workspace.owner_id in self.user_workspaces:
                self.user_workspaces[workspace.owner_id].discard(workspace_id)

            # Remove from in-memory cache
            if workspace_id in self.active_workspaces:
                del self.active_workspaces[workspace_id]

            # Remove from Redis cache
            await redis_cache.delete(key=self._get_cache_key(workspace_id))

            # Delete from database
            async with get_session() as db:
                db_workspace = await db.get(WorkspaceModel, str(workspace_id))

                if db_workspace:
                    await db.delete(db_workspace)
                    await db.commit()
                else:
                    logger.warning(f"Workspace not found in database: {workspace_id}")

            logger.info(f"Deleted workspace: {workspace_id}")
            return True

        except Exception as e:
            logger.error(f"Failed to delete workspace: {str(e)}", exc_info=True)
            return False

    async def archive_workspace(self, workspace_id: uuid.UUID) -> bool:
        """
        Archive a workspace

        Args:
            workspace_id: Workspace ID to archive

        Returns:
            True if archived, False if not found or error
        """
        try:
            # Update workspace with archived status
            workspace = await self.update_workspace(
                workspace_id=workspace_id,
                is_archived=True,
            )

            return workspace is not None

        except Exception as e:
            logger.error(f"Failed to archive workspace: {str(e)}", exc_info=True)
            return False

    async def get_user_workspaces(
        self,
        user_id: uuid.UUID,
        include_archived: bool = False,
    ) -> List[Workspace]:
        """
        Get all workspaces for a user

        Args:
            user_id: User ID
            include_archived: Whether to include archived workspaces

        Returns:
            List of workspaces
        """
        try:
            result = []

            # Check in-memory mapping
            workspace_ids = self.user_workspaces.get(user_id, set())

            if workspace_ids:
                # Get workspaces from cache or database
                for workspace_id in workspace_ids:
                    workspace = await self.get_workspace(workspace_id)

                    if workspace:
                        # Filter archived workspaces if needed
                        if include_archived or not workspace.is_archived:
                            result.append(workspace)

            # If no workspaces found in memory, query database
            if not result:
                async with get_session() as db:
                    query = db.query(WorkspaceModel).filter(
                        WorkspaceModel.owner_id == str(user_id)
                    )

                    if not include_archived:
                        query = query.filter(WorkspaceModel.is_archived == False)

                    db_workspaces = await query.all()

                    for db_workspace in db_workspaces:
                        # Convert to Workspace model
                        workspace = Workspace(
                            id=uuid.UUID(db_workspace.id),
                            owner_id=uuid.UUID(db_workspace.owner_id)
                            if db_workspace.owner_id
                            else None,
                            created_at=db_workspace.created_at,
                            updated_at=db_workspace.updated_at,
                            last_active_at=db_workspace.last_active_at,
                            is_archived=db_workspace.is_archived,
                            is_private=db_workspace.is_private,
                            metadata=WorkspaceMetadata(**db_workspace.metadata),
                        )

                        # Add to caches
                        workspace_id = workspace.id
                        self.active_workspaces[workspace_id] = workspace
                        await self._cache_workspace(workspace)

                        # Update user workspace mapping
                        if user_id not in self.user_workspaces:
                            self.user_workspaces[user_id] = set()
                        self.user_workspaces[user_id].add(workspace_id)

                        result.append(workspace)

            return result

        except Exception as e:
            logger.error(f"Failed to get user workspaces: {str(e)}", exc_info=True)
            return []

    def _get_cache_key(self, workspace_id: uuid.UUID) -> str:
        """Get Redis cache key for a workspace"""
        return f"{self.WORKSPACE_CACHE_PREFIX}{workspace_id}"

    async def _cache_workspace(self, workspace: Workspace) -> None:
        """Cache a workspace in Redis"""
        key = self._get_cache_key(workspace.id)

        await redis_cache.set(
            key=key,
            value=workspace.dict(),
            ttl=self.WORKSPACE_CACHE_TTL,
        )


# Global instance
workspace_manager = None


def initialize_workspace_manager() -> WorkspaceManager:
    """
    Initialize the global workspace manager instance

    Returns:
        The initialized workspace manager
    """
    global workspace_manager
    if workspace_manager is None:
        workspace_manager = WorkspaceManager()
    return workspace_manager


# Export public symbols
__all__ = [
    "WorkspaceMetadata",
    "Workspace",
    "WorkspaceManager",
    "workspace_manager",
    "initialize_workspace_manager",
]

File: cortex/node/src/components/workspace-manager.ts
/**
 * Implementation of the Workspace Manager component for Cortex Core
 */

import { v4 as uuidv4 } from 'uuid';
import { logger } from '../utils/logger';
import { db } from '../database/connection';
import { redisClient } from '../cache/redis';
import { stringifyJson, parseJsonString } from '../utils/json-helpers';

export interface Workspace {
  id: string;
  userId: string;
  name: string;
  createdAt: Date;
  lastActiveAt: Date;
  config: WorkspaceConfig;
  metadata: Record<string, any>;
}

export interface WorkspaceConfig {
  defaultModality?: string;
  sharingEnabled: boolean;
  retentionDays: number;
}

export interface Conversation {
  id: string;
  workspaceId: string;
  modality: string;
  title: string;
  createdAt: Date;
  lastActiveAt: Date;
  entries: ConversationEntry[];
  metadata: Record<string, any>;
}

export interface ConversationEntry {
  id: string;
  type: "user" | "assistant" | "system";
  content: any;
  timestamp: Date;
  metadata?: Record<string, any>;
}

export interface ConversationFilter {
  modality?: string;
  fromDate?: Date;
  toDate?: Date;
  searchText?: string;
}

// Define database record types for proper typing
interface WorkspaceRecord {
  id: string;
  userId: string;
  name: string;
  createdAt: Date;
  lastActiveAt: Date;
  config: string; // JSON string
  metadata: string; // JSON string
}

interface ConversationRecord {
  id: string;
  workspaceId: string;
  modality: string;
  title: string;
  createdAt: Date;
  lastActiveAt: Date;
  entries: string; // JSON string
  metadata: string; // JSON string
}

export class WorkspaceManager {
  private readonly WORKSPACE_CACHE_PREFIX = 'workspace:';
  private readonly CONVERSATION_CACHE_PREFIX = 'conversation:';
  private readonly CACHE_TTL = 3600; // 1 hour in seconds

  /**
   * Create a new workspace
   * @param userId The ID of the user
   * @param name The name of the workspace
   * @param config Optional workspace configuration
   * @returns The newly created workspace
   */
  async createWorkspace(userId: string, name: string, config?: Partial<WorkspaceConfig>): Promise<Workspace> {
    try {
      logger.info(`Creating workspace "${name}" for user ${userId}`);

      const defaultConfig: WorkspaceConfig = {
        defaultModality: config?.defaultModality || 'chat',
        sharingEnabled: config?.sharingEnabled ?? false,
        retentionDays: config?.retentionDays || 90
      };

      const now = new Date();
      const workspaceId = uuidv4();

      // Create workspace in database with JSON serialization
      const workspace = await db.workspaces.create({
        data: {
          id: workspaceId,
          userId,
          name,
          createdAt: now,
          lastActiveAt: now,
          config: stringifyJson(defaultConfig),  // Convert to JSON string
          metadata: stringifyJson({})            // Convert to JSON string
        }
      });

      // Parse the JSON strings to create the properly typed workspace object
      const typedWorkspace: Workspace = {
        id: workspace.id,
        userId: workspace.userId,
        name: workspace.name,
        createdAt: workspace.createdAt,
        lastActiveAt: workspace.lastActiveAt,
        config: parseJsonString(workspace.config, defaultConfig),
        metadata: parseJsonString(workspace.metadata, {})
      };

      // Cache workspace for faster access
      await this.cacheWorkspace(typedWorkspace);

      logger.info(`Created workspace ${workspaceId} for user ${userId}`);
      return typedWorkspace;
    } catch (error) {
      logger.error(`Failed to create workspace for user ${userId}`, error);
      throw new Error(`Failed to create workspace: ${(error as Error).message}`);
    }
  }

  /**
   * Get workspace by ID
   * @param workspaceId The ID of the workspace
   * @returns The workspace or null if not found
   */
  async getWorkspace(workspaceId: string): Promise<Workspace | null> {
    try {
      logger.debug(`Getting workspace ${workspaceId}`);

      // Try to get from cache first
      const cachedWorkspace = await redisClient.get(`${this.WORKSPACE_CACHE_PREFIX}${workspaceId}`);
      if (cachedWorkspace) {
        return JSON.parse(cachedWorkspace) as Workspace;
      }

      // If not in cache, get from database
      const workspace = await db.workspaces.findUnique({
        where: { id: workspaceId }
      });

      if (!workspace) {
        return null;
      }

      // Parse JSON strings into objects
      const typedWorkspace: Workspace = {
        id: workspace.id,
        userId: workspace.userId,
        name: workspace.name,
        createdAt: workspace.createdAt,
        lastActiveAt: workspace.lastActiveAt,
        config: parseJsonString(workspace.config, {
          defaultModality: 'chat',
          sharingEnabled: false,
          retentionDays: 90
        }),
        metadata: parseJsonString(workspace.metadata, {})
      };

      // Cache for future access
      await this.cacheWorkspace(typedWorkspace);

      return typedWorkspace;
    } catch (error) {
      logger.error(`Failed to get workspace ${workspaceId}`, error);
      return null;
    }
  }

  /**
   * List workspaces for a user
   * @param userId The ID of the user
   * @returns Array of workspaces
   */
  async listWorkspaces(userId: string): Promise<Workspace[]> {
    try {
      logger.debug(`Listing workspaces for user ${userId}`);

      // Get workspaces from database
      const workspaces = await db.workspaces.findMany({
        where: { userId }
      });

      // Parse JSON strings and cache workspaces
      const typedWorkspaces: Workspace[] = workspaces.map((workspace: WorkspaceRecord) => {
        const typedWorkspace: Workspace = {
          id: workspace.id,
          userId: workspace.userId,
          name: workspace.name,
          createdAt: workspace.createdAt,
          lastActiveAt: workspace.lastActiveAt,
          config: parseJsonString(workspace.config, {
            defaultModality: 'chat',
            sharingEnabled: false,
            retentionDays: 90
          }),
          metadata: parseJsonString(workspace.metadata, {})
        };

        // Cache each workspace
        this.cacheWorkspace(typedWorkspace).catch(err => {
          logger.error(`Failed to cache workspace ${workspace.id}`, err);
        });

        return typedWorkspace;
      });

      return typedWorkspaces;
    } catch (error) {
      logger.error(`Failed to list workspaces for user ${userId}`, error);
      return [];
    }
  }

  /**
   * Create a conversation in a workspace
   * @param workspaceId The ID of the workspace
   * @param modality The modality of the conversation
   * @param title Optional title for the conversation
   * @returns The newly created conversation
   */
  async createConversation(workspaceId: string, modality: string, title?: string): Promise<Conversation> {
    try {
      logger.info(`Creating ${modality} conversation in workspace ${workspaceId}`);

      const workspace = await this.getWorkspace(workspaceId);
      if (!workspace) {
        throw new Error(`Workspace not found: ${workspaceId}`);
      }

      const now = new Date();
      const conversationId = uuidv4();

      // Generate a title if none provided
      const conversationTitle = title || `${modality.charAt(0).toUpperCase() + modality.slice(1)} conversation ${now.toLocaleDateString()}`;

      // Create conversation in database with JSON serialization
      const conversation = await db.conversations.create({
        data: {
          id: conversationId,
          workspaceId,
          modality,
          title: conversationTitle,
          createdAt: now,
          lastActiveAt: now,
          entries: stringifyJson([]),     // Empty array as JSON string
          metadata: stringifyJson({})     // Empty object as JSON string
        }
      });

      // Parse JSON strings into objects
      const typedConversation: Conversation = {
        id: conversation.id,
        workspaceId: conversation.workspaceId,
        modality: conversation.modality,
        title: conversation.title,
        createdAt: conversation.createdAt,
        lastActiveAt: conversation.lastActiveAt,
        entries: parseJsonString(conversation.entries, []),
        metadata: parseJsonString(conversation.metadata, {})
      };

      // Cache conversation for faster access
      await this.cacheConversation(typedConversation);

      // Update workspace last active time
      await this.updateWorkspaceLastActive(workspaceId);

      logger.info(`Created conversation ${conversationId} in workspace ${workspaceId}`);
      return typedConversation;
    } catch (error) {
      logger.error(`Failed to create conversation in workspace ${workspaceId}`, error);
      throw new Error(`Failed to create conversation: ${(error as Error).message}`);
    }
  }

  /**
   * Get a conversation by ID
   * @param conversationId The ID of the conversation
   * @returns The conversation or null if not found
   */
  async getConversation(conversationId: string): Promise<Conversation | null> {
    try {
      logger.debug(`Getting conversation ${conversationId}`);

      // Try to get from cache first
      const cachedConversation = await redisClient.get(`${this.CONVERSATION_CACHE_PREFIX}${conversationId}`);
      if (cachedConversation) {
        return JSON.parse(cachedConversation) as Conversation;
      }

      // If not in cache, get from database
      const conversation = await db.conversations.findUnique({
        where: { id: conversationId }
      });

      if (!conversation) {
        return null;
      }

      // Parse JSON strings into objects
      const typedConversation: Conversation = {
        id: conversation.id,
        workspaceId: conversation.workspaceId,
        modality: conversation.modality,
        title: conversation.title,
        createdAt: conversation.createdAt,
        lastActiveAt: conversation.lastActiveAt,
        entries: parseJsonString(conversation.entries, []),
        metadata: parseJsonString(conversation.metadata, {})
      };

      // Cache for future access
      await this.cacheConversation(typedConversation);

      return typedConversation;
    } catch (error) {
      logger.error(`Failed to get conversation ${conversationId}`, error);
      return null;
    }
  }

  /**
   * List conversations in a workspace
   * @param workspaceId The ID of the workspace
   * @param filter Optional filter for conversations
   * @returns Array of conversations
   */
  async listConversations(workspaceId: string, filter?: ConversationFilter): Promise<Conversation[]> {
    try {
      logger.debug(`Listing conversations for workspace ${workspaceId}`);

      // Build filter conditions for database query
      const where: any = { workspaceId };

      if (filter?.modality) {
        where.modality = filter.modality;
      }

      if (filter?.fromDate) {
        where.createdAt = { gte: filter.fromDate };
      }

      if (filter?.toDate) {
        where.createdAt = {
          ...where.createdAt,
          lte: filter.toDate
        };
      }

      // Get conversations from database
      let conversationsDb = await db.conversations.findMany({ where });

      // Parse JSON strings into objects
      const typedConversations: Conversation[] = conversationsDb.map((conversation: ConversationRecord) => {
        const typedConversation: Conversation = {
          id: conversation.id,
          workspaceId: conversation.workspaceId,
          modality: conversation.modality,
          title: conversation.title,
          createdAt: conversation.createdAt,
          lastActiveAt: conversation.lastActiveAt,
          entries: parseJsonString(conversation.entries, []),
          metadata: parseJsonString(conversation.metadata, {})
        };

        return typedConversation;
      });

      // Apply text search filter if specified
      let filteredConversations = typedConversations;
      if (filter?.searchText && filter.searchText.trim().length > 0) {
        const searchLower = filter.searchText.toLowerCase();
        filteredConversations = typedConversations.filter(conv =>
          conv.title.toLowerCase().includes(searchLower) ||
          JSON.stringify(conv.metadata).toLowerCase().includes(searchLower)
        );
      }

      // Cache all conversations for faster access later
      for (const conversation of filteredConversations) {
        await this.cacheConversation(conversation);
      }

      return filteredConversations;
    } catch (error) {
      logger.error(`Failed to list conversations for workspace ${workspaceId}`, error);
      return [];
    }
  }

  /**
   * Add an entry to a conversation
   * @param conversationId The ID of the conversation
   * @param entry The entry to add
   */
  async addConversationEntry(conversationId: string, entry: Omit<ConversationEntry, 'id'>): Promise<ConversationEntry> {
    try {
      logger.info(`Adding entry to conversation ${conversationId}`);

      const conversation = await this.getConversation(conversationId);
      if (!conversation) {
        throw new Error(`Conversation not found: ${conversationId}`);
      }

      // Generate ID for entry if not provided
      const entryWithId: ConversationEntry = {
        id: uuidv4(),
        ...entry
      };

      // Add entry to conversation
      conversation.entries.push(entryWithId);
      conversation.lastActiveAt = new Date();

      // Update conversation in database
      await db.conversations.update({
        where: { id: conversationId },
        data: {
          entries: stringifyJson(conversation.entries),    // Convert array to JSON string
          lastActiveAt: conversation.lastActiveAt
        }
      });

      // Update cache
      await this.cacheConversation(conversation);

      // Update workspace last active time
      await this.updateWorkspaceLastActive(conversation.workspaceId);

      return entryWithId;
    } catch (error) {
      logger.error(`Failed to add entry to conversation ${conversationId}`, error);
      throw new Error(`Failed to add conversation entry: ${(error as Error).message}`);
    }
  }

  /**
   * Update workspace last active time
   * @param workspaceId The ID of the workspace
   * @private
   */
  private async updateWorkspaceLastActive(workspaceId: string): Promise<void> {
    try {
      const now = new Date();

      // Update in database
      await db.workspaces.update({
        where: { id: workspaceId },
        data: { lastActiveAt: now }
      });

      // Update in cache if exists
      const cachedWorkspace = await redisClient.get(`${this.WORKSPACE_CACHE_PREFIX}${workspaceId}`);
      if (cachedWorkspace) {
        const workspace = JSON.parse(cachedWorkspace) as Workspace;
        workspace.lastActiveAt = now;
        await this.cacheWorkspace(workspace);
      }
    } catch (error) {
      logger.error(`Failed to update workspace last active time: ${workspaceId}`, error);
    }
  }

  /**
   * Cache a workspace for fast access
   * @param workspace The workspace to cache
   * @private
   */
  private async cacheWorkspace(workspace: Workspace): Promise<void> {
    const cacheKey = `${this.WORKSPACE_CACHE_PREFIX}${workspace.id}`;

    await redisClient.set(
      cacheKey,
      JSON.stringify(workspace),
      'EX',
      this.CACHE_TTL
    );
  }

  /**
   * Cache a conversation for fast access
   * @param conversation The conversation to cache
   * @private
   */
  private async cacheConversation(conversation: Conversation): Promise<void> {
    const cacheKey = `${this.CONVERSATION_CACHE_PREFIX}${conversation.id}`;

    await redisClient.set(
      cacheKey,
      JSON.stringify(conversation),
      'EX',
      this.CACHE_TTL
    );
  }
}
